### PR TITLE
Fixes #21287 - add resource switcher

### PR DESCRIPTION
**Pull request by amirfefer**

### CODE REVIEW

**Comment by waldenraines** [comment]
@amirfefer how do you feel about deploying the storybook so I can have a look more easily?

**Comment by amirfefer** [comment]
@waldenraines  sure :+1:

File: webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/ResourceSwitcher.stories.js
Comment (timogoebel) [inline webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/ResourceSwitcher.stories.js]: Nit: should be `switcher`

**Comment by amirfefer** [comment]
@waldenraines 
Storybook link: http://rawgit.com/amirfefer/foreman/resource-switcher/index.html?selectedKind=Resource%20Switcher&selectedStory=swithcer&full=0&down=1&left=1&panelRight=0&downPanel=storybook%2Factions%2Factions-panel

**Comment by amirfefer** [comment]
@sharvit  could you have a look?

File: webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/ResourceSwitcher.css
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/ResourceSwitcher.css]: Would be nice to use variables here, so plugins could override them instead of overriding the class itself.

File: webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/ResurceSwitcher.test.js
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/ResurceSwitcher.test.js]: Really need it here? Shouldn't we allow jest to mock all the imports?

File: webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js]: Would it make sense to support both `href` and `onClick`?

File: webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js]: Can you add `propTypes` please?

**Review by sharvit — CHANGES_REQUESTED** [review]
Overall, it feels good.

1. Is it still WIP? if so, what is actually missing here?

2. You can use the `onOutsideClick` to allow auto toggling:
https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/notifications/index.js#L80

3. The design doesn't look like the patternfly design.

**Comment by amirfefer** [comment]
@sharvit 
Thanks for your review!
I've removed `WIP` tag from this PR.
I addressed your inline comments plus,
Instead of using `onOutsideClick` as you suggested, I used `rootClose` prop (already implemented in react-bootstrap)

File: webpack/assets/javascripts/react_app/components/bookmarks/__snapshots__/bookmarks.test.js.snap
Comment (waldenraines) [inline webpack/assets/javascripts/react_app/components/bookmarks/__snapshots__/bookmarks.test.js.snap]: Nitpick, seems accidental.

File: webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js
Comment (waldenraines) [inline webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js]: What if we just omit the title?  If you look at the version in katello it doesn't have a title but instead has a search bar.

screen shot 2018-02-20 at 2 25 44 pm

**Review by sharvit — CHANGES_REQUESTED** [review]
LGTM except for the filename changing.

Do you have to change the filename?

This pr #5212 is changing the filename same as you do:
https://github.com/theforeman/foreman/pull/5212/files#diff-c81edf5233381be438a6cd30882c7c19R1
https://github.com/theforeman/foreman/pull/5212/files#diff-f394b153b2a3a9f4f039a8743b649144R1

I would prefer you will keep it the same filename so it will be easier for me to rebase.

**Comment by amirfefer** [comment]
@waldenraines Thanks, I removed that title.
@sharvit  - I changed the file name back.

**Review by sharvit — APPROVED** [review]
LGTM!
@ohadlevy

**Comment by amirfefer** [comment]
[test katello]

**Comment by amirfefer** [comment]
@ohadlevy could we merge this?

**Review by waldenraines — COMMENTED** [review]
I hate to bring this up so late but what do we think about calling this "Item Switcher" instead of "Resource Switcher"?  I'm not opposed to "Resource Switcher" but "Item Switcher" may be more generic and is shorter.

Otherwise this looks good to me.

**Review by dLobatog — COMMENTED** [review]
Looks good to me too, just 2 questions:

- I see the Katello one is not a button but a clickable link, should this be the same?
- What page is this going to be used on?

File: webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/ResourceSwitcher.stories.js
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/ResourceSwitcher.stories.js]: just wondering, why not use the fixtures? they seem very close?

File: webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js]: what if this is a very long list? should we limit it somehow? did you guys discuss on if we should filter it?

File: webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js]: just wondering, as i didn't fully catched up, is this structure in line with @sharvit proposal ?

File: webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js]: yes we should use an autocomplete search bar, as soon as `autocomplete` component will merge, I'll update this. With this iterative approach I think we can merge components faster

**Comment by amirfefer** [comment]
@dLobatog Thanks for your comment.
I followed [patternfly design](http://www.patternfly.org/pattern-library/navigation/breadcrumbs/#design), Patternfly uses a button for it, as you can see here:
item-switcher
and it should combined with the breadcrumbs, in order to switch to a different item - another way to navigate. 

@waldenraines I don't mind to replace the name, `item switcher` sounds even better for me.

File: webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js]: no... it's not a redux connected component.

**Comment by waldenraines** [comment]
> I see the Katello one is not a button but a clickable link, should this be the same?

~~@Rohoover thoughts on this?  I can update the katello side if we decide on a button.~~

ETA: never mind, I didn't realize that this is now in patternfly.  I will update the katello side to use a button.

File: webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js
Comment (waldenraines) [reply webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js]: The original mockups for the katello side include search and pagination of the results.

**Comment by waldenraines** [comment]
> @waldenraines I don't mind to replace the name, item switcher sounds even better for me.

Maybe we should call it breadcrumb switcher to match the name of the patternfly design?  I didn't realize that this had been put into patternfly.  In that case I wonder if it makes sense to move this to patternfly-react.

File: webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js]: @waldenraines 
I don't like the pagination here, looks odd in a dropdown menu, patternfly design doesn't include it either.

**Review by Rohoover — COMMENTED** [review]
@waldenraines 

It was in process in PF when you implemented. Your implementation I think, quickened its contribution :)
 Sorry I should have notified you.

File: webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js
Comment (waldenraines) [reply webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js]: @Rohoover thoughts?

Comment (Rohoover) [reply webpack/assets/javascripts/react_app/components/common/ResourceSwitcher/index.js]: I included pagination because of the large number of rows. Forcing the user to search for something doesn't allow for browsing. I also did not want to re-introduce anything like infinite scroll since previous customer feedback was disastrous.

**Comment by amirfefer** [comment]
@waldenraines, I renamed it to `BreadcrumbSwitcher`
@ohadlevy - I reuse `fixtures` file in the storybook story file.

**Comment by amirfefer** [comment]
[test foreman]

**Comment by waldenraines** [comment]
@amirfefer and what are your thoughts on going ahead and moving this into patternfly-react?

**Comment by amirfefer** [comment]
@waldenraines I totally agree on that, though auto complete component still WIP, so it would take some time until this get merge there and the time quite pressing. we can do these both - merge this into foreman, and open a PR in pf-react ?

**Comment by waldenraines** [comment]
> @waldenraines I totally agree on that, though auto complete component still WIP, so it would take some time until this get merge there and the time quite pressing. we can do these both - merge this into foreman, and open a PR in pf-react ?

Works for me, thanks @amirfefer

**Comment by amirfefer** [comment]
@ohadlevy 
could we merge this?

File: webpack/assets/javascripts/react_app/components/common/BreadcrumbSwitcher/BreadcrumbSwitcher.test.js
Comment (waldenraines) [inline webpack/assets/javascripts/react_app/components/common/BreadcrumbSwitcher/BreadcrumbSwitcher.test.js]: Doesn't this need to be updated to BreadcrumbSwitcher?

**Review by waldenraines — CHANGES_REQUESTED** [review]

**Comment by ohadlevy** [comment]
@amirfefer can you rebase please? tests are currently failing.

File: .storybook/config.js
Comment (ohadlevy) [inline .storybook/config.js]: are you mocking all window objects?? did you mean to do this instead: https://github.com/theforeman/foreman/pull/5019/files#diff-a7e25ac69ff4b688eb41c907225d770bR19

File: webpack/assets/javascripts/react_app/components/common/BreadcrumbSwitcher/index.js
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/common/BreadcrumbSwitcher/index.js:25]: will this work with turbolinks?

**Review by ohadlevy — COMMENTED** [review]
a couple of in line questions, but a more generic question is why isnt this part of pf-react instead of foreman to begin with? thanks!

**Comment by ohadlevy** [comment]
@amirfefer also a rebase is needed... please make sure your PR are green....

File: webpack/assets/javascripts/react_app/components/common/BreadcrumbSwitcher/index.js
Comment (amirfefer) [reply webpack/assets/javascripts/react_app/components/common/BreadcrumbSwitcher/index.js:25]: yes, `ListGroupItem` renders `<a>` tag when href exists.

**Comment by amirfefer** [comment]
@ohadlevy, @waldenraines   thanks
fixed and rebased.

**Comment by waldenraines** [comment]
> a couple of in line questions, but a more generic question is why isnt this part of pf-react instead of foreman to begin with? thanks!

As @amirfefer detailed above, we need this for 6.4 and there is some concern that it won't get into patternfly-react in a timely manner.  We fully intend to move this to patternfly-react very soon though.

**Comment by waldenraines** [comment]
@amirfefer looks like there are additional changes needed here to get this green.

**Comment by amirfefer** [comment]
@waldenraines  tests seems unrelated

File: .storybook/preview-head.html
Comment (ohadlevy) [inline .storybook/preview-head.html:14]: 1. why do we need storybook code to be aware of jest? vs a mock in a test?
2. this is pure JS (doesnt go via webpack) hence you might shouldnt use ES6.

**Comment by ohadlevy** [comment]
[test foreman]

**Comment by ohadlevy** [comment]
[test katello]

**Comment by ohadlevy** [comment]
[test upgrade]

**Review by ohadlevy — COMMENTED** [review]
I would be happy to see a PR to patternfly soon, I really can't see a reason why not to merge it to PF

File: .storybook/preview-head.html
Comment (amirfefer) [reply .storybook/preview-head.html:14]: because i'm using fixtures file in order to be more dry.
https://github.com/theforeman/foreman/pull/5188/files#diff-7f590775f94b18cdd0bc39583a9a7d40R6

**Comment by waldenraines** [comment]
> I would be happy to see a PR to patternfly soon, I really can't see a reason why not to merge it to PF

If I recall correctly it's because we are waiting on finalized design from patternfly.  @Rohoover is there a finalized design for the breadcrumb switcher yet?

**Comment by Rohoover** [comment]
@waldenraines 
I believe so... it can't be found here:

https://github.com/Rohoover/patternfly-design/blob/master/pattern-library/navigation/breadcrumbs/design/design.md

Is that what you're looking for?

**Comment by waldenraines** [comment]
> @waldenraines I totally agree on that, though auto complete component still WIP, so it would take some time until this get merge there and the time quite pressing. we can do these both - merge this into foreman, and open a PR in pf-react ?

@amirfefer the autocomplete component is one that I know was waiting on a finalized design (see https://github.com/patternfly/patternfly-react/pull/190#issuecomment-364109632) and that was the reason for the delay.  I think we should try getting this into patternfly-react and if there are problems doing so we can move this back into foreman temporarily.

**Comment by waldenraines** [comment]
> @amirfefer the autocomplete component is one that I know was waiting on a finalized design (see patternfly/patternfly-react#190 (comment)) and that was the reason for the delay. I think we should try getting this into patternfly-react and if there are problems doing so we can move this back into foreman temporarily.

@amirfefer thoughts on this?  Honestly I've seen patternfly-react be much quicker than getting foreman PRs merged 😄

**Comment by ohadlevy** [comment]
thats probably my fault, if you can help me identify which PR's are tested
/ ready to be merged I'll be happy to have another look

**Comment by waldenraines** [comment]
> thats probably my fault, if you can help me identify which PR's are tested
> / ready to be merged I'll be happy to have another look

Not your fault just an observation.  It's really the small number of people with commit access that is the problem.  We are working on helping to alleviate this on our team.

**Comment by waldenraines** [comment]
@amirfefer was this closed because we are moving this to patternfly-react?

**Pull request closed without merging**
